src/briefcase/integrations/subprocess.py

```diff
@@ -244,26 +296,47 @@ def stream_output(self, label, popen_process):
         :param label: A description of the content being streamed; used for
             to provide context in logging messages.
         :param popen_process: a running Popen process with output to print
+        :param stop_func: a Callable that returns True when output streaming
+            should stop and the popen_process should be terminated.
         """
+        output_streamer = threading.Thread(
+            name=f"{label} output streamer",
+            target=self._stream_output_thread,
+            args=(popen_process,),
+        )
         try:
-            while True:
-                # readline should always return at least a newline (ie \n)
-                # UNLESS the underlying process is exiting/gone; then "" is returned
-                output_line = ensure_str(popen_process.stdout.readline())
-                if output_line:
-                    self.command.logger.info(output_line)
-                elif output_line == "":
-                    # a return code will be available once the process returns one to the OS.
-                    # by definition, that should mean the process has exited.
-                    return_code = popen_process.poll()
-                    # only return once all output has been read and the process has exited.
-                    if return_code is not None:
-                        self._log_return_code(return_code)
-                        return
-
+            output_streamer.start()
+            if stop_func:
+                while output_streamer.is_alive():
+                    if stop_func():
+                        self.cleanup(label, popen_process)
+                    time.sleep(0.5)
+            else:
+                output_streamer.join()
         except KeyboardInterrupt:
             self.cleanup(label, popen_process)
 
+    def _stream_output_thread(self, popen_process):
```

Comment:
If I'm reading this right, this is still a useful function to have around - looking forward to #629, we might even end up using it with the Windows launcher. I'd suggest keeping this "as is" as `stream_output_thread`, and modify the other method to be an "stream_output_thread_background" that has the termination callable.

Reply:
Gotcha. I'll leave this as internal to `Subprocess` then for now.